Improved Reference Metric and SIMD

ksible opened the pull request:
I simplified the algorithm for reference_metric.py and added backward-compatibility since SymPy changed the derivative representation with SymPy 1.2. I implemented an expression tree data structure (SIMDExprTree.py) for direct manipulation of the SymPy expression tree with better performance than the build-in xreplace function. I improved the performance of the SIMD algorithm (SIMD.py) for NRPy using the previously mentioned expression tree.

zachetienne commented on SIMD.py:
```diff
@@ -222,89 +239,80 @@ def IntegerPowSIMD(a, n):
-
-        if item != tempitem: expr = expr.subs(item, tempitem)
-
-    # Step 7: SIMD intrinsics cannot take integers or rational numbers as arguments.
```
I think you chopped off this comment. Please add it back, revise if need be.

zachetienne commented on reference_metric.py:
```diff
@@ -626,38 +626,23 @@ def ref_metric__hatted_quantities(SymPySimplifyExpressions=True):
+                        var, order = str(item.args[1])[2:], len(item.args) - 1
+                    else:
+                        var, order = str(item.args[1][0])[2:], item.args[1][1]
+                    oper = '__D' + 'D'*(order - 1) + var*order
```
Please describe what is being done here. Provide an example input & output.

zachetienne commented on SIMDExprTree.py:
```diff
@@ -0,0 +1,134 @@
+__author__ = 'Ken Sible'
```
Please add a header consistent with other NRPy+ Python modules that contains author info and briefly describes the purpose of the module.

ksible commented:
I fixed the issue with Travis-CI failing on Python 2. The error that kept arising was "only named arguments may follow *expression," which occurs whenever a non-keyword argument follows tuple unpacking. I removed the tuple unpacking from that failed section of the SIMD script. Furthermore, the string interpolation in SIMDExprTree was replaced with string formatting for compatibility.
